### ISSUE TITLE
Failed to set Win32 parent window of the Control.

### ISSUE THREAD

**Issue by DanLBarr** [dotnet/winforms#5669]
* .NET Core Version: 
.NET 6 Preview 7

* Have you experienced this same bug with .NET Framework?: 
No. This problem also does NOT occur with .NET 5

**Problem description:**
We have a System.Windows.Forms.TabControl with multiple tabs in a Form. Each tab contains a custom control of ours which inherits from System.Windows.Forms.UserControl. The problem occurs when our custom control loads a graphic model which creates Windows.Forms controls using reflection. The first tab works fine. However, when you click on another tab, a crash occurs. Following is the crash call stack.

System.ComponentModel.Win32Exception
  HResult=0x80004005
  Message=Failed to set Win32 parent window of the Control.
  Source=System.Windows.Forms
  StackTrace:
   at System.Windows.Forms.Control.SetParentHandle(IntPtr value)
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.SetVisibleCore(Boolean value)
   at System.Windows.Forms.TabControl.UpdateTabSelection(Boolean updateFocus)
   at System.Windows.Forms.TabControl.OnSelectedIndexChanged(EventArgs e)
   at System.Windows.Forms.TabControl.WmSelChange()
   at System.Windows.Forms.TabControl.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.SendMessageW(IntPtr hWnd, WM Msg, IntPtr wParam, IntPtr lParam)
   at System.Windows.Forms.Control.ReflectMessage(IntPtr hWnd, Message& m)
   at System.Windows.Forms.Control.WmNotify(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.CallWindowProcW(IntPtr wndProc, IntPtr hWnd, WM msg, IntPtr wParam, IntPtr lParam)
   at System.Windows.Forms.NativeWindow.DefWndProc(Message& m)
   at System.Windows.Forms.Control.WmMouseDown(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.TabControl.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop(UIntPtr dwComponentID, msoloop uReason, Void* pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(msoloop reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(msoloop reason, ApplicationContext context)
   at System.Windows.Forms.Application.Run(Form mainForm)
   at EmbeddedControls.Program.Main() in C:\slaves\VS2022\m70_slave.netX\demo.netX\EmbeddedControls\EmbeddedControls\Program.cs:line 19

**Expected behavior:**
It shouldn't crash.

**Minimal repro:**
I am working on reproducing the problem without requiring our custom control library (which is very complex).

**Comment by dreddy-work** [comment]
@DanLBarr, thanks for looking into a simple repro here. Please provide your test sample once you have it. By any chance, did you set  any High DPI mode in the application?

**Comment by DanLBarr** [comment]
@dreddy-work Thank you for your comment. No, high DPI mode is not set. Application.EnableVisualStyles() is also not called. We do call Application.SetCompatibleTextRendererDefault(false).

**Comment by dreddy-work** [comment]
Okay. Thank you for confirming. In this case, We would need a simple repro app for us to investigate further.

**Comment by DanLBarr** [comment]
@dreddy-work I've discovered this issue is not related to the System.Windows.Forms controls that our model can contain and create. I've created a scenario with two TabControls where each tab contains an instance of our UserControl and neither of our controls have a model loaded. That seems to indicate it must have to do with the UserControl, not it's contents. If I remove one of our UserControls from one of the tabs, the problem goes away. In the exception stack I see a Control is being created and it looks like SetParentHandle() is setting the parent of the new Control? What is that newly created Control and how does that relate to our UserControl?

In our UserControl we override CreateParams() and set the ClassStyle to CS_OWNDC. We also override WndProc() and process WM_INPUTLANGCHANGE, WM_CTLCOLORBTN, WM_CTLCOLOREDIT, WM_CTLCOLORLISTBOX, WM_CTLCOLORSCROLLBAR, WM_CTLCOLORSTATIC, WM_HSCROLL, WM_VSCROLL and WM_DRAWITEM. After we are done, we pass the message on to base.WndProc() except for the WM_DRAWITEM. Would any of this be related? What is the tab control doing in the exception stack we see?

At this point I'm not sure how much of the functionality I would need to include in a test to reproduce this problem.

Thanks

**Comment by dreddy-work** [comment]
@DanLBarr, Code path shows a tab selection changed causing the controls in the selected `Tabpage `to be created. Unfortunately, Native error code here is very generic. We expect a failure in this case only if you end up with different DPIs between child and parent somehow, but you ruled that out. Did you try same scenario with a simple user control?  Given that this is a regression between .NET 5.0 and 6.0, we can prioritize investigation. We need a consistent repro steps. @Olina-Zhang , can you see if you guys can repro this scenario?

**Comment by DanLBarr** [comment]
@dreddy-work Yes, I tried the same scenario with a plain vanilla UserControl implementation. It did not crash. You can find the project attached.
[WinFormsApp2.zip](https://github.com/dotnet/winforms/files/7102915/WinFormsApp2.zip)

**Comment by Olina-Zhang** [comment]
@DanLBarr Yes, cannot repro with Winforms user control in TabPages on form. So could you please provide a sample application for this issue?

**Comment by DanLBarr** [comment]
@Olina-Zhang Thank you, I am working on being able to provide a sample which will reproduce the problem. I have taken a simple UserControl and am following the exception call stack to see what might be triggering the exception. Based on that, I am adding code to the UserControl to see if I can reproduce the problem.

**Comment by dreddy-work** [comment]
@DanLBarr, any luck here?

**Comment by DanLBarr** [comment]
@dreddy-work No luck yet. Our user control is complex and includes native code interfaced to .NET by a C++/CLI layer. Is there a way we could securely send you a project with our DLLs so you can reproduce it? It would also include a temporary license for our library.

**Comment by RussKie** [comment]
You can raise a VS Feedback (upper right hand side in VS) and then attach all necessary files to that. Once raise please post the link to the ticket here so we make sure it gets assigned to our team.

**Comment by DanLBarr** [comment]
@RussKie @dreddy-work I will do that tomorrow when our release manager can put together the package. They are on east coast time. Thank you

**Comment by DanLBarr** [comment]
@RussKie @dreddy-work I've reported the problem through the Developer Community and uploaded the files for reproducing the problem.

**Comment by RussKie** [comment]
@DanLBarr unfortunately the link you posted isn't linking to your issue, it links to your feedback start page.

**Comment by DanLBarr** [comment]
@RussKie sorry, I just copied the link from the page I had up. I think this is the proper link. The description is the same as the GitHub issue.

https://developercommunity.visualstudio.com/t/Failed-to-set-Win32-parent-window-of-the/1526052?entry=myfeedback&space=61&q=%22Failed+to+set+Win32+parent+window+of+the+Control%22

**Comment by RussKie** [comment]
@DanLBarr got it 👍 

@dreddy-work assigned to you

**Comment by DanLBarr** [comment]
@RussKie I thought that the attachment to the Developer Community would be private. It seems to be downloadable by anyone.

**Comment by dreddy-work** [comment]
> @RussKie I thought that the attachment to the Developer Community would be private. It seems to be downloadable by anyone.

@DanLBarr , sorry for the confusion. There is an option to share as private in the comments section but not in the original content of the feedback. For now, We took the linked attachment down. We have the sample for repro'ing the issue and update what we find here.

**Comment by dreddy-work** [comment]
@DanLBarr , I was able to run the app first time and see the exception. However, when i try running it again, it silently quits.  Any cache that needs to be cleared here?

**Comment by DanLBarr** [comment]
@dreddy-work No, this application doesn't have any cache. I just tried it on my machine and I get the exception every time (both running the executable directly from the command line and from Visual Studio 2022). I did find when opening the project in Visual Studio you will need to copy the SL.GMS*.dll files to the bin directory since that is where they were originally referenced in the project.

**Comment by dreddy-work** [comment]
May be something specific to my machine. Any steps to run it via Visual Studio?  I see this: 
![Issue1](https://user-images.githubusercontent.com/36968667/133170585-a0a41ce0-a342-4ba9-ab66-4b5c9f909e25.gif)

**Comment by DanLBarr** [comment]
@dreddy-work That almost looks like it isn't finding a license. After you execute gms_init.bat, if you type "set gms" at the command prompt it should show the following "GMS_HOME=C:\Users\dreddy\Downloads\test\SL-5669_RBT\DB_TEST". There should also be a file named KEYS in the lib directory. If you open KEYS with something like Notepad, the content of KEYS should be the text
00aa4665005f8d48 # temporary key

**Comment by dreddy-work** [comment]
> @dreddy-work That almost looks like it isn't finding a license. After you execute gms_init.bat, if you type "set gms" at the command prompt it should show the following: GMS_HOME=C:\Users\dreddy\Downloads\test\SL-5669_RBT\DB_TEST. There should also be a file named KEYS in the lib directory. If you open KEYS with something like Notepad, the content of KEYS should be the text
> 00aa4665005f8d48 # temporary key

Yes, i see them but application still exits silently ( i see two processes launched)

**Comment by DanLBarr** [comment]
@dreddy-work That is strange! I only see one process. 
![task_manager](https://user-images.githubusercontent.com/87153506/133176682-4c067d31-c66a-414c-9651-d407723a0064.png)
If you want to build the program with Visual Studio, do devenv from your command prompt so it gets the environment. Open the solution, build and run it.

**Comment by dreddy-work** [comment]
@DanLBarr , Thanks for the details. I see what is wrong here. You were trying to set parent for a SystemAware control  from a Unaware control. Please below gif.  List Viewer from library is running in SystemAware while your application (TabPage) running in `UnAware `mode.  Adding `Application.SetHighDpiMode(HighDpiMode.SystemAware);` in `program.cs `in the sample app you shared should unblock you.

![Issue2](https://user-images.githubusercontent.com/36968667/133181669-a02dd358-cf11-449c-adcc-a647e86932d8.gif)

Closing this issue as it is specific to you application and you can easily fix it.

**Comment by RussKie** [comment]
@dreddy-work I propose we reword the exception message to provide more information to a developer.

E.g.
```cs
if (User32.SetParent(.... == 0) 
{
    var controlDpi = User32.GetAwarenessFromDpiAwarenessContext(...(Handle));
    var parentDpi = User32.GetAwarenessFromDpiAwarenessContext(...(value));
    if (controlDpi != parentDpi)
    {
        throw new Win32Exception("DPI don't match");
    }
    else
    {
        throw new Win32Exception(<current wording>);
    }
}
```

**Comment by DanLBarr** [comment]
@dreddy-work Thank you! That fixed the immediate problem and now works in the full demo program. I may have some questions tomorrow about how this will affect our control.

**Comment by DanLBarr** [comment]
@dreddy-work Wondering why the first tab doesn't have the same problem?

Another question is what could set DPI awareness on our UserControl? We actually would want that but I am not aware of anything we've done to set it. I've searched for "dpi" in all of our source code and didn't find anything.

As for the exception message, it would be nice to have something like "The parent/child DPI awareness don't match" but it would also be helpful to know which controls are involved.

Thanks,
Daniel
